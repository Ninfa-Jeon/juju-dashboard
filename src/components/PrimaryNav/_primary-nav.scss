@import "vanilla-framework/scss/settings_colors";
@import "../../scss/settings";

.p-primary-nav {
  position: sticky;
  top: 0;
  width: $sidebar-full-width;

  a,
  a:hover {
    color: $color-light;
    text-decoration: none;
  }

  &__logo {
    max-width: 110px;
    padding: 0.5rem 0.5rem 0;
    width: 100%;
  }

  &__header {
    background: $color-navigation-dark;
    display: grid;
    grid-template-columns: 1fr 40px;
    padding: 0.5rem 0.5rem 0;
  }

  &__toggle {
    background: none;
    border: 0;
    margin: 0;
    padding: 0;
    transition: transform 0.25s ease-in-out;

    &:hover {
      background: none;
    }
  }

  .is-external {
    background: $color-navigation-dark;
    margin-bottom: 0.5rem;
    max-height: 0;
    overflow: hidden;
    transition: max-height 0.25s ease-in-out;

    .p-list__link {
      padding-left: calc(3.5rem - 3px);
    }
  }

  .p-list__item {
    padding: 0;

    .p-list__link {
      border-left: 3px solid transparent;
      color: $color-light;
      display: flex;
      padding: 0.25rem 0;
      padding-left: calc(1.25rem - 3px);

      &:hover {
        background-color: $color-navigation-highlight;
      }

      &.is-selected {
        background-color: $color-navigation-highlight;
        border-left-color: $color-navigation-accent;
      }
    }

    .p-list__icon {
      height: 16px;
      padding-right: 1.25rem;
      position: relative;
      top: 0.2rem;
    }
  }

  &.ext-nav-open {
    .is-external {
      max-height: 200px;
      padding-bottom: 0.5rem;
    }

    .p-primary-nav__toggle {
      transform: rotate(180deg);
    }
  }

  .entity-count {
    background-color: $color-negative;
    border-radius: 50%;
    display: inline-block;
    font-size: 0.875rem;
    height: 1.5rem;
    margin-left: auto;
    margin-right: 1rem;
    padding: 0 0.25rem;
    text-align: center;
    width: 1.5rem;
  }
}
