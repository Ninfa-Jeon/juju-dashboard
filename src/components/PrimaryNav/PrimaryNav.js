import React, { useState } from "react";
import { Link } from "react-router-dom";
import classNames from "classnames";

import "./_primary-nav.scss";

const pages = [
  { label: "Models", path: "/" },
  { label: "Controllers", path: "/controllers" },
  { label: "Usage", path: "/usage" },
  { label: "Logs", path: "/logs" }
];

const PrimaryNav = () => {
  const currentLocation = window.location.pathname;
  const [extNavOpen, setExtNavOpen] = useState(false);
  return (
    <nav
      className={classNames("p-primary-nav", { "ext-nav-open": extNavOpen })}
    >
      <div className="p-primary-nav__header">
        <Link to="/">
          <img
            className="p-primary-nav__logo"
            src="https://assets.ubuntu.com/v1/a9e0ed4a-jaas-logo1.svg"
            alt="JAAS logo"
          />
        </Link>
        <button
          className="p-primary-nav__toggle"
          onClick={() => setExtNavOpen(!extNavOpen)}
        >
          <i className="p-icon--contextual-menu">Toggle external navigation</i>
        </button>
      </div>
      <ul className="p-list is-external">
        <li className="p-list__item">
          <a
            href="https://jaas.ai/store"
            target="_blank"
            rel="noopener noreferrer"
          >
            Store
          </a>
        </li>
        <li className="p-list__item">
          <a
            href="https://jaas.ai/about"
            target="_blank"
            rel="noopener noreferrer"
          >
            About
          </a>
        </li>
        <li className="p-list__item">
          <a
            href="https://jaas.ai/how-it-works"
            target="_blank"
            rel="noopener noreferrer"
          >
            How it works
          </a>
        </li>
        <li className="p-list__item">
          <a
            className="p-link--external"
            href="https://discourse.jujucharms.com/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Discourse
          </a>
        </li>
        <li className="p-list__item">
          <a className="p-link--external" href="https://jaas.ai/docs/">
            Docs
          </a>
        </li>
      </ul>
      <ul className="p-list is-internal">
        {pages.map(navItem => (
          <li
            key={navItem.path}
            className={classNames("p-list__item", {
              "is-selected": currentLocation === navItem.path
            })}
          >
            <Link className="p-list__link" to={navItem.path}>
              {navItem.label}
            </Link>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default PrimaryNav;
