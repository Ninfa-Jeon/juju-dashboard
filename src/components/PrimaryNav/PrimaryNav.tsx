import { StatusLabel } from "@canonical/react-components";
import classNames from "classnames";
import Logo from "components/Logo/Logo";
import UserMenu from "components/UserMenu/UserMenu";
import { useSelector } from "react-redux";
import { NavLink } from "react-router-dom";
import { getAppVersion } from "store/general/selectors";
import {
  getControllersUpdateCount,
  getGroupedModelStatusCounts,
} from "store/juju/selectors";
import "./_primary-nav.scss";

const ModelsLink = () => {
  const { blocked: blockedModels } = useSelector(getGroupedModelStatusCounts);
  return (
    <NavLink
      className={({ isActive }) =>
        classNames("p-list__link", {
          "is-selected": isActive,
        })
      }
      to={"/models"}
    >
      <i className={`p-icon--models is-light`}></i>
      <span className="hide-collapsed">Models</span>
      {blockedModels > 0 && (
        <span className="entity-count is-negative">{blockedModels}</span>
      )}
    </NavLink>
  );
};

const ControllersLink = () => {
  const controllersUpdateCount = useSelector(getControllersUpdateCount());
  return (
    <NavLink
      className={({ isActive }) =>
        classNames("p-list__link", {
          "is-selected": isActive,
        })
      }
      to={"/controllers"}
    >
      <i className={`p-icon--controllers is-light`}></i>
      <span className="hide-collapsed">Controllers</span>
      {controllersUpdateCount > 0 && (
        <span className="entity-count is-caution">
          {controllersUpdateCount}
        </span>
      )}
    </NavLink>
  );
};

const PrimaryNav = () => {
  const appVersion = useSelector(getAppVersion);
  return (
    <nav className="p-primary-nav">
      <div className="p-primary-nav__header">
        <Logo />
      </div>

      <ul className="p-list is-internal">
        <li className="p-list__item">
          <ModelsLink />
        </li>
        <li className="p-list__item">
          <ControllersLink />
        </li>
      </ul>
      <hr className="p-primary-nav__divider hide-collapsed" />
      <div className="p-primary-nav__bottom hide-collapsed">
        <ul className="p-list">
          <li className="p-list__item">
            <a
              className="p-list__link"
              href="https://github.com/canonical-web-and-design/jaas-dashboard/issues/new"
              target="_blank"
              rel="noopener noreferrer"
            >
              Report a bug
            </a>
          </li>
        </ul>
      </div>
      <hr className="p-primary-nav__divider hide-collapsed" />
      <div className="p-primary-nav__bottom hide-collapsed">
        <ul className="p-list">
          <li className="p-list__item">
            <span className="version">Version {appVersion}</span>
            <StatusLabel appearance="positive">Beta</StatusLabel>
          </li>
        </ul>
      </div>
      <UserMenu />
    </nav>
  );
};

export default PrimaryNav;
