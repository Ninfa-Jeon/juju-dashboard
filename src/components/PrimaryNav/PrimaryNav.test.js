import React from "react";
import { BrowserRouter as Router, MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import dataDump from "testing/complete-redux-store-dump";

import PrimaryNav from "./PrimaryNav";

const mockStore = configureStore([]);
describe("Primary Nav", () => {
  it("renders without crashing and matches snapshot", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <PrimaryNav />
        </Router>
      </Provider>
    );
    expect(wrapper.find(".p-primary-nav")).toMatchSnapshot();
  });

  it("toggles external nav menu", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <PrimaryNav />
        </Router>
      </Provider>
    );

    const primaryNav = ".p-primary-nav";
    const primaryNavToggle = ".p-primary-nav__toggle";

    expect(wrapper.find(primaryNav).hasClass("ext-nav-open")).toEqual(false);
    wrapper.find(primaryNavToggle).simulate("click");
    expect(wrapper.find(primaryNav).hasClass("ext-nav-open")).toEqual(true);
    wrapper.find(primaryNavToggle).simulate("click");
    expect(wrapper.find(primaryNav).hasClass("ext-nav-open")).toEqual(false);
  });

  it("applies is-selected state correctly", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/logs"]}>
          <PrimaryNav />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("a.is-selected").text()).toStrictEqual("Logs");
  });

  it("displays correct number of blocked models", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/"]}>
          <PrimaryNav />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".entity-count").text()).toStrictEqual("2");
  });
});
