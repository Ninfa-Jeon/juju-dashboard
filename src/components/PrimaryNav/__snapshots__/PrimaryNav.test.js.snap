// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primary Nav renders without crashing and matches snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <PrimaryNav>
      <nav
        className="p-primary-nav"
      >
        <div
          className="p-primary-nav__header"
        >
          <Link
            to="/"
          >
            <LinkAnchor
              href="/"
              navigate={[Function]}
            >
              <a
                href="/"
                onClick={[Function]}
              >
                <img
                  alt="JAAS logo"
                  className="p-primary-nav__logo"
                  src="https://assets.ubuntu.com/v1/a9e0ed4a-jaas-logo1.svg"
                />
              </a>
            </LinkAnchor>
          </Link>
          <button
            className="p-primary-nav__toggle"
            onClick={[Function]}
          >
            <i
              className="p-icon--contextual-menu"
            >
              Toggle external navigation
            </i>
          </button>
        </div>
        <ul
          className="p-list is-external"
        >
          <li
            className="p-list__item"
          >
            <a
              className="p-list__link"
              href="https://jaas.ai/store"
              rel="noopener noreferrer"
              target="_blank"
            >
              Store
            </a>
          </li>
          <li
            className="p-list__item"
          >
            <a
              className="p-list__link"
              href="https://jaas.ai/about"
              rel="noopener noreferrer"
              target="_blank"
            >
              About
            </a>
          </li>
          <li
            className="p-list__item"
          >
            <a
              className="p-list__link"
              href="https://jaas.ai/how-it-works"
              rel="noopener noreferrer"
              target="_blank"
            >
              How it works
            </a>
          </li>
          <li
            className="p-list__item"
          >
            <a
              className="p-link--external p-list__link"
              href="https://discourse.jujucharms.com/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Discourse
            </a>
          </li>
          <li
            className="p-list__item"
          >
            <a
              className="p-link--external p-list__link"
              href="https://jaas.ai/docs/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Docs
            </a>
          </li>
        </ul>
        <ul
          className="p-list is-internal"
        >
          <li
            className="p-list__item"
            key="/"
          >
            <NavLink
              activeClassName="is-selected"
              className="p-list__link"
              exact={true}
              to="/"
            >
              <Link
                aria-current="page"
                className="p-list__link is-selected"
                style={Object {}}
                to={
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": null,
                  }
                }
              >
                <LinkAnchor
                  aria-current="page"
                  className="p-list__link is-selected"
                  href="/"
                  navigate={[Function]}
                  style={Object {}}
                >
                  <a
                    aria-current="page"
                    className="p-list__link is-selected"
                    href="/"
                    onClick={[Function]}
                    style={Object {}}
                  >
                    <img
                      alt="Models icon"
                      className="p-list__icon"
                      src="https://assets.ubuntu.com/v1/51642f58-models.svg"
                    />
                    Models
                  </a>
                </LinkAnchor>
              </Link>
            </NavLink>
          </li>
          <li
            className="p-list__item"
            key="/controllers"
          >
            <NavLink
              activeClassName="is-selected"
              className="p-list__link"
              exact={true}
              to="/controllers"
            >
              <Link
                aria-current={null}
                className="p-list__link"
                to={
                  Object {
                    "hash": "",
                    "pathname": "/controllers",
                    "search": "",
                    "state": null,
                  }
                }
              >
                <LinkAnchor
                  aria-current={null}
                  className="p-list__link"
                  href="/controllers"
                  navigate={[Function]}
                >
                  <a
                    aria-current={null}
                    className="p-list__link"
                    href="/controllers"
                    onClick={[Function]}
                  >
                    <img
                      alt="Controllers icon"
                      className="p-list__icon"
                      src="https://assets.ubuntu.com/v1/8414b187-controllers.svg"
                    />
                    Controllers
                  </a>
                </LinkAnchor>
              </Link>
            </NavLink>
          </li>
          <li
            className="p-list__item"
            key="/usage"
          >
            <NavLink
              activeClassName="is-selected"
              className="p-list__link"
              exact={true}
              to="/usage"
            >
              <Link
                aria-current={null}
                className="p-list__link"
                to={
                  Object {
                    "hash": "",
                    "pathname": "/usage",
                    "search": "",
                    "state": null,
                  }
                }
              >
                <LinkAnchor
                  aria-current={null}
                  className="p-list__link"
                  href="/usage"
                  navigate={[Function]}
                >
                  <a
                    aria-current={null}
                    className="p-list__link"
                    href="/usage"
                    onClick={[Function]}
                  >
                    <img
                      alt="Usage icon"
                      className="p-list__icon"
                      src="https://assets.ubuntu.com/v1/1b1d07ae-usage.svg"
                    />
                    Usage
                  </a>
                </LinkAnchor>
              </Link>
            </NavLink>
          </li>
          <li
            className="p-list__item"
            key="/logs"
          >
            <NavLink
              activeClassName="is-selected"
              className="p-list__link"
              exact={true}
              to="/logs"
            >
              <Link
                aria-current={null}
                className="p-list__link"
                to={
                  Object {
                    "hash": "",
                    "pathname": "/logs",
                    "search": "",
                    "state": null,
                  }
                }
              >
                <LinkAnchor
                  aria-current={null}
                  className="p-list__link"
                  href="/logs"
                  navigate={[Function]}
                >
                  <a
                    aria-current={null}
                    className="p-list__link"
                    href="/logs"
                    onClick={[Function]}
                  >
                    <img
                      alt="Logs icon"
                      className="p-list__icon"
                      src="https://assets.ubuntu.com/v1/741097ff-logs.svg"
                    />
                    Logs
                  </a>
                </LinkAnchor>
              </Link>
            </NavLink>
          </li>
        </ul>
      </nav>
    </PrimaryNav>
  </Router>
</BrowserRouter>
`;
